app/routes/user/email_notifications.server.ts

```diff
@@ -0,0 +1,208 @@
+ * of the National Aeronautics and Space Administration. No copyright is claimed
+ * in the United States under Title 17, U.S. Code. All Other Rights Reserved.
+ *
+ * SPDX-License-Identifier: NASA-1.3
+ */
+import { tables } from '@architect/functions'
+import { mapTopicsToFormatAndNoticeType } from '~/lib/utils'
+import { getUser } from '~/routes/__auth/user.server'
+
+// db model
+export type EmailNotification = {
+  name: string
+  recipient: string
+  created: number
+  active: boolean
+  uuid?: string
+  topics: string[]
+}
+
+// view model
+export type EmailNotificationVM = EmailNotification & {
+  format: string
+  noticeTypes: string[]
+}
+
+export class EmailNotificationVendingMachine {
+  #sub: string
+
+  private constructor(sub: string) {
+    this.#sub = sub
+  }
+
+  // Init machine
+  static async create(request: Request) {
+    const user = await getUser(request)
+    if (!user) throw new Response('not signed in', { status: 403 })
+    return new this(user.sub)
+  }
+
+  // Create
+  async createEmailNotification(notification: EmailNotification) {
+    if (!notification.name)
+      throw new Response('name must not be empty', { status: 400 })
+    if (!notification.recipient)
+      throw new Response('recipient must not be empty', { status: 400 })
+    if (!notification.topics)
+      throw new Response('topics must not be empty', { status: 400 })
+
+    const created = Date.now()
+    const uuid = crypto.randomUUID()
+
+    const db = await tables()
+    await db.email_notification.put({
+      sub: this.#sub,
+      uuid,
+      name: notification.name,
+      created,
+      topics: notification.topics,
+      active: true,
+      recipient: notification.recipient,
+    })
+
+    // Materialized view
+    for (const topic of notification.topics) {
+      await db.email_notification_subscription.put({
+        uuid,
+        topic,
+        recipient: notification.recipient,
+      })
+    }
+  }
+
+  // Read
+  async getEmailNotifications() {
+    const db = await tables()
+    const results = await db.email_notification.query({
+      KeyConditionExpression: '#sub = :sub',
+      ExpressionAttributeNames: {
+        '#sub': 'sub',
+        '#uuid': 'uuid',
+        '#name': 'name',
+        '#created': 'created',
+        '#topics': 'topics',
+        '#recipient': 'recipient',
+      },
+      ExpressionAttributeValues: {
+        ':sub': this.#sub,
+      },
+      ProjectionExpression: '#uuid, #created, #name, #topics, #recipient',
+    })
+    const emailNotifications = results.Items as EmailNotificationVM[]
+    for (const notice of emailNotifications) {
+      let formats = []
+      notice.noticeTypes = []
+      for (const topic of notice.topics) {
+        let mappedData = mapTopicsToFormatAndNoticeType(topic)
+        formats.push(mappedData.noticeFormat)
+        notice.noticeTypes.push(mappedData.noticeType)
+      }
+      notice.format = [...new Set(formats)][0]
+    }
```

Comment:
I don't know about you, but I prefer `array.map(...)` to `for (... of ...)` because it avoids having to dynamically grow the result and it lets the TypeScript compiler do more type inference for you.